File: leetcode/Python/house-robber-iii.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def rob(self, root: Optional[TreeNode]) -> int:
        um = [dict(), dict()]
        def traverse(node, visit, num):
            if node is None:
                return 0
            r = um[visit].get(num, None)
            if r is not None:
                return r
            if visit==1:
                return node.val + traverse(node.left,0,2*num) + traverse(node.right,0,2*num+1)
            a, b= traverse(node.left, 1, 2*num), traverse(node.left, 0, 2*num)
            c, d = traverse(node.right, 1, 2*num+1), traverse(node.right, 0, 2*num+1)
            r=um[visit][num]=max(a,b)+max(c,d)
            return r
        return max(traverse(root,1,1), traverse(root,0,1))

File: leetcode/C++/house-robber-iii.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    unordered_map<long,int> um[2];
    unordered_map<long,int>::iterator it;
    int traverse(TreeNode* node, int visit, long num){
        if(node==nullptr) return 0;
        if((it=um[visit].find(num))!=um[visit].end()) return it->second;
        if(visit) return node->val + traverse(node->left, 0, 2*num) + traverse(node->right, 0, 2*num+1);
        int a=traverse(node->left,1,2*num), b=traverse(node->left,0,2*num);
        int c=traverse(node->right,1,2*num+1), d=traverse(node->right,0,2*num+1);
        return um[visit][num]=max(a,b)+max(c,d);
    }
    int rob(TreeNode* root) {
        return max(traverse(root,1, 1), traverse(root,0,1));
    }
};